File: legacy-dotnet/Execution/RetryPolicy.cs
using System.Collections.Concurrent;

namespace AdoAutopilot.Execution;

public class RetryState
{
    public int RetryCount { get; set; }
    public DateTime LastAttempt { get; set; }
    public string? LastError { get; set; }
}

public class RetryPolicy
{
    private readonly ConcurrentDictionary<int, RetryState> _states = new();
    private readonly int _maxRetries;
    private readonly int _backoffSeconds;
    private readonly ILogger<RetryPolicy> _logger;

    public RetryPolicy(int maxRetries, int backoffSeconds, ILogger<RetryPolicy> logger)
    {
        _maxRetries = maxRetries;
        _backoffSeconds = backoffSeconds;
        _logger = logger;
    }

    public bool ShouldRetry(int workItemId) =>
        !_states.TryGetValue(workItemId, out var s) || s.RetryCount < _maxRetries;

    public bool IsBackoffActive(int workItemId)
    {
        if (!_states.TryGetValue(workItemId, out var s)) return false;
        var delay = TimeSpan.FromSeconds(_backoffSeconds * Math.Pow(2, s.RetryCount - 1));
        return DateTime.UtcNow - s.LastAttempt < delay;
    }

    public void RecordFailure(int workItemId, string error)
    {
        _states.AddOrUpdate(workItemId,
            _ => new RetryState { RetryCount = 1, LastAttempt = DateTime.UtcNow, LastError = error },
            (_, existing) =>
            {
                existing.RetryCount++;
                existing.LastAttempt = DateTime.UtcNow;
                existing.LastError = error;
                return existing;
            });

        var state = _states[workItemId];
        _logger.LogWarning("#{Id} failed (attempt {Count}/{Max}): {Error}",
            workItemId, state.RetryCount, _maxRetries, error);
    }

    public void RecordSuccess(int workItemId) => _states.TryRemove(workItemId, out _);

    public bool IsExhausted(int workItemId) =>
        _states.TryGetValue(workItemId, out var s) && s.RetryCount >= _maxRetries;

    public RetryState? GetState(int workItemId) =>
        _states.TryGetValue(workItemId, out var s) ? s : null;
}

File: ai_autopilot/execution/retry_policy.py
"""Per-work-item retry tracking with exponential backoff (ported from ``RetryPolicy``)."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import UTC, datetime

from ai_autopilot.logging_config import get_logger


@dataclass
class RetryState:
    retry_count: int = 0
    last_attempt: datetime = field(default_factory=lambda: datetime.now(UTC))
    last_error: str | None = None


class RetryPolicy:
    def __init__(self, max_retries: int, backoff_seconds: int) -> None:
        self._max_retries = max_retries
        self._backoff_seconds = backoff_seconds
        self._states: dict[int, RetryState] = {}
        self._log = get_logger("execution.retry_policy")

    def should_retry(self, work_item_id: int) -> bool:
        state = self._states.get(work_item_id)
        return state is None or state.retry_count < self._max_retries

    def is_backoff_active(self, work_item_id: int) -> bool:
        state = self._states.get(work_item_id)
        if state is None:
            return False
        delay = self._backoff_seconds * (2 ** (state.retry_count - 1))
        elapsed = (datetime.now(UTC) - state.last_attempt).total_seconds()
        return elapsed < delay

    def record_failure(self, work_item_id: int, error: str) -> None:
        state = self._states.get(work_item_id)
        if state is None:
            state = RetryState(retry_count=1, last_error=error)
            self._states[work_item_id] = state
        else:
            state.retry_count += 1
            state.last_attempt = datetime.now(UTC)
            state.last_error = error
        self._log.warning(
            "work item failed",
            id=work_item_id,
            attempt=state.retry_count,
            max=self._max_retries,
            error=error,
        )

    def record_success(self, work_item_id: int) -> None:
        self._states.pop(work_item_id, None)

    def is_exhausted(self, work_item_id: int) -> bool:
        state = self._states.get(work_item_id)
        return state is not None and state.retry_count >= self._max_retries

    def get_state(self, work_item_id: int) -> RetryState | None:
        return self._states.get(work_item_id)
